pkg/reconciler/controllers/remediator.go

```diff
@@ -0,0 +1,343 @@
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog/v2"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/declared"
+	"kpt.dev/configsync/pkg/diff"
+	"kpt.dev/configsync/pkg/importer/analyzer/validation/nonhierarchical"
+	"kpt.dev/configsync/pkg/metadata"
+	"kpt.dev/configsync/pkg/metrics"
+	"kpt.dev/configsync/pkg/remediator"
+	"kpt.dev/configsync/pkg/remediator/cache"
+	"kpt.dev/configsync/pkg/status"
+	syncerclient "kpt.dev/configsync/pkg/syncer/client"
+	syncerreconcile "kpt.dev/configsync/pkg/syncer/reconcile"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Remediator is a controller that watches for timer-based events.
+// When an object is added to the remediate resource list, it enqueues the GKNN
+// of the object as a remediate request. The reconciler dequeues the request
+// and remediate the object to match the declared resource.
+type Remediator struct {
+	// ControllerCtx is set in reconciler/main.go and is canceled by the Finalizer.
+	ControllerCtx context.Context
+	// DoneCh indicates the Remediator stops processing new requests.
+	// It is a signal for the Finalizer to continue to finalize managed resources.
+	DoneCh chan struct{}
+	// SyncScope is the scope of the reconciler, either a namespace or ':root'
+	SyncScope declared.Scope
+	// SyncName is the name of the RootSync or RepoSync.
+	SyncName string
+	// Applier is the client to correct drift.
+	syncerreconcile.Applier
+	// declared.Resources is the cache of the declared resource list. It is used
+	// as the source of truth for drift correction.
+	*declared.Resources
+	// cache.RemediateResources is the cache of the remediate resource list. It is
+	// used to get the actual Kubernetes object for a particular GKNN.
+	*cache.RemediateResources
+	// Remediator is the actor to check whether it is ready to start reconciliation.
+	*remediator.Remediator
+}
+
+// Reconcile processes the remediate request to correct drift.
+func (r *Remediator) Reconcile(_ context.Context, req ctrl.Request) (ctrl.Result, error) {
+	if errors.Is(r.ControllerCtx.Err(), context.Canceled) {
+		// Drain the request queue
+		klog.V(3).Infof("Stop remediating new objects as the context is canceled..")
+		return ctrl.Result{}, nil
+	}
+
+	klog.V(1).Infof("New remediator reconciliation for %q", req.Name)
+
+	id, err := core.FromID(req.Name)
+	if err != nil {
+		klog.Errorf("invalid request: %v", err)
+		return ctrl.Result{}, nil
+	}
+
+	if !r.CanRemediate() {
+		// Mark the object as done so that it can be requeued later when the Applier is not running.
+		r.DoneObj(id)
+		return ctrl.Result{}, nil
+	}
+
+	// Mark the Remediator is working to block the Syncer.
+	r.Remediating()
```

Comment:
They are not atomic because `Remediating` doesn't always run after `CanRemediate`.
`CanRemediate` is invoked in two places, while `Remediating` is invoked only once:
- When registering the Remediator in `SetupWithManager`, we started a goroutine to enqueue the request when `CanRemediate`, but `Remediating` isn't invoked here.
- In the `reconcile` function, it calls `Remediating` when `CanRemediate` is true.

If we make them atomic, we need to allow enqueuing when the applier is running. I think that is okay as one object only gets updated once in the apply, and will be enqueued after the apply completes. Even though it is enqueued earlier, it won't get processed until the apply completes.

Reply:
After a second thought, we still need to check `CanRemediate` when enqueueing requests. Otherwise, it will keep attempting to enqueue while the applier is running.
The enqueue attempts are unnecessary for long-running apply.

We can still have two functions, one check if it can remediate only, and the other does both check and remediate atomically.

Reply:
What's important here is that the `canRemediate` check and the `waitgroup.Add` call happen in the same critical section.

If we need a separate function signature which doesn't call `waitgroup.Add` to support the enqueue logic that's fine.

For example, change the `Remediating` function signature to:
```golang
// Remediating indicates the Remediator starts processing the remediate requests.
func (r *Remediator) Remediating() bool {
	r.lifecycleMux.Lock()
	defer r.lifecycleMux.Unlock()

	klog.V(1).Info("Remediating...")
    if r.canRemediate {
        r.wg.Add(1)
    }
    return r.canRemediate
}
```

Reply:
done